File: Classes/SBTextInputView.m
//
//  SBTextInputView.m
//  Sunrise
//
//  Created by Atsushi Jike on 09/11/28.
//  Copyright 2010 Atsushi Jike. All rights reserved.
//

#import "SBTextInputView.h"


@implementation SBTextInputView

@dynamic message;
@dynamic text;

- (id)initWithFrame:(NSRect)frame prompt:(NSString *)prompt
{
	if (self = [super initWithFrame:frame])
	{
		[self constructMessageLabel:prompt];
		[self constructTextLabel];
		[self setAutoresizingMask:(NSViewMinXMargin | NSViewMaxXMargin | NSViewMinYMargin | NSViewMaxYMargin)];
	}
	return self;
}

- (void)dealloc
{
	[messageLabel release];
	[textLabel release];
	[doneButton release];
	[cancelButton release];
	[super dealloc];
}

#pragma mark Rects

- (NSPoint)margin
{
	return NSMakePoint(36.0, 32.0);
}

- (CGFloat)labelWidth
{
	return 85.0;
}

- (NSSize)buttonSize
{
	return NSMakeSize(105.0, 24.0);
}

- (CGFloat)buttonMargin
{
	return 15.0;
}

- (NSFont *)textFont
{
	return [NSFont systemFontOfSize:16];
}

- (NSRect)messageLabelRect
{
	NSRect r = NSZeroRect;
	NSPoint margin = [self margin];
	r.size.width = self.bounds.size.width - margin.x * 2;
	r.size.height = 36.0;
	r.origin.x = margin.x;
	r.origin.y = self.bounds.size.height - r.size.height - margin.y;
	return r;
}

- (NSRect)textLabelRect
{
	NSRect r = NSZeroRect;
	NSRect messageLabelRect = [self messageLabelRect];
	NSPoint margin = [self margin];
	NSSize ButtonSize = [self buttonSize];
	r.size.width = self.bounds.size.width - margin.x * 2;
	r.size.height = (self.bounds.size.height - messageLabelRect.size.height - ButtonSize.height) - margin.y * 4;
	r.origin.x = margin.x;
	r.origin.y = messageLabelRect.origin.y - r.size.height;
	return r;
}

- (NSRect)doneButtonRect
{
	NSRect r = NSZeroRect;
	NSPoint margin = [self margin];
	CGFloat buttonMargin = [self buttonMargin];
	r.size = [self buttonSize];
	r.origin.y = margin.y;
	r.origin.x = (self.bounds.size.width - (r.size.width * 2 + buttonMargin)) / 2 + r.size.width + buttonMargin;
	return r;
}

- (NSRect)cancelButtonRect
{
	NSRect r = NSZeroRect;
	NSPoint margin = [self margin];
	CGFloat buttonMargin = [self buttonMargin];
	r.size = [self buttonSize];
	r.origin.y = margin.y;
	r.origin.x = (self.bounds.size.width - (r.size.width * 2 + buttonMargin)) / 2;
	return r;
}

#pragma mark Construction

- (void)constructMessageLabel:(NSString *)inMessage
{
	NSRect r = [self messageLabelRect];
	NSFont *font = [NSFont boldSystemFontOfSize:16];
	NSSize size = [inMessage sizeWithAttributes:[NSDictionary dictionaryWithObject:font forKey:NSFontAttributeName]];
	messageLabel = [[NSTextField alloc] initWithFrame:r];
	[messageLabel setAutoresizingMask:(NSViewMinXMargin | NSViewMinYMargin)];
	[messageLabel setEditable:NO];
	[messageLabel setBordered:NO];
	[messageLabel setDrawsBackground:NO];
	[messageLabel setTextColor:[NSColor whiteColor]];
	[[messageLabel cell] setFont:font];
	[[textLabel cell] setAlignment:size.width > (r.size.width - 20.0) ? NSLeftTextAlignment : NSCenterTextAlignment];
	[[messageLabel cell] setWraps:YES];
	[messageLabel setStringValue:inMessage];
	[self addSubview:messageLabel];
}

- (void)constructTextLabel
{
	NSRect r = [self textLabelRect];
	textLabel = [[SBBLKGUITextField alloc] initWithFrame:r];
	[textLabel setAlignment:NSLeftTextAlignment];
	[textLabel setFont:[NSFont systemFontOfSize:14.0]];
	[textLabel setTextColor:[NSColor whiteColor]];
	[textLabel setDelegate:self];
	[[textLabel cell] setWraps:YES];
	[self addSubview:textLabel];
}

- (void)constructDoneButton
{
	NSRect r = [self doneButtonRect];
	doneButton = [[SBBLKGUIButton alloc] initWithFrame:r];
	[doneButton setTitle:NSLocalizedString(@"OK", nil)];
	[doneButton setTarget:self];
	[doneButton setAction:@selector(done)];
	[doneButton setEnabled:YES];
	[doneButton setKeyEquivalent:@"\r"];	// busy if button is added into a view
	[self addSubview:doneButton];
}

- (void)constructCancelButton
{
	NSRect r = [self cancelButtonRect];
	cancelButton = [[SBBLKGUIButton alloc] initWithFrame:r];
	[cancelButton setTitle:NSLocalizedString(@"Cancel", nil)];
	[cancelButton setTarget:self];
	[cancelButton setAction:@selector(cancel)];
	[cancelButton setKeyEquivalent:@"\e"];
	[self addSubview:cancelButton];
}

#pragma mark Delegate

- (void)controlTextDidChange:(NSNotification *)aNotification
{
	BOOL canDone = NO;
	canDone = [[textLabel stringValue] length] > 0;
	[doneButton setEnabled:canDone];
}

#pragma mark Getter

- (NSString *)message
{
	return [messageLabel stringValue];
}

- (NSString *)text
{
	return [textLabel stringValue];
}

#pragma mark Setter

- (void)setMessage:(NSString *)message
{
	[messageLabel setStringValue:message];
}

- (void)setText:(NSString *)inText
{
	[textLabel setStringValue:inText];
}

- (void)setDoneSelector:(SEL)inDoneSelector
{
	if (doneSelector != inDoneSelector)
	{
		doneSelector = inDoneSelector;
		if (doneSelector && !doneButton)
		{
			[self constructDoneButton];
		}
	}
}

- (void)setCancelSelector:(SEL)inCancelSelector
{
	if (cancelSelector != inCancelSelector)
	{
		cancelSelector = inCancelSelector;
		if (cancelSelector && !cancelButton)
		{
			[self constructCancelButton];
		}
	}
}

@end

File: Classes/SBTextInputView.swift
/*
SBTextInputView.swift

Copyright (c) 2014, Alice Atlas
Copyright (c) 2010, Atsushi Jike
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright notice,
this list of conditions and the following disclaimer in the documentation
and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

import BLKGUI

class SBTextInputView: SBView, NSTextFieldDelegate {
    lazy var messageLabel: NSTextField = {
        let font = NSFont.boldSystemFontOfSize(16)
        let messageLabel = NSTextField(frame: self.messageLabelRect)
        messageLabel.autoresizingMask = [.ViewMinXMargin, .ViewMinYMargin]
        messageLabel.editable = false
        messageLabel.bordered = false
        messageLabel.drawsBackground = false
        messageLabel.textColor = NSColor.whiteColor()
        messageLabel.font = font
        messageLabel.cell!.wraps = true
        return messageLabel
    }()
    lazy var textLabel: BLKGUI.TextField = {
        let textLabel = BLKGUI.TextField(frame: self.textLabelRect)
        textLabel.alignment = .Left
        textLabel.font = NSFont.systemFontOfSize(14.0)
        textLabel.textColor = NSColor.whiteColor()
        textLabel.delegate = self
        textLabel.cell!.wraps = true
        return textLabel
    }()
    lazy var doneButton: BLKGUI.Button = {
        let doneButton = BLKGUI.Button(frame: self.doneButtonRect)
        doneButton.title = NSLocalizedString("OK", comment: "")
        doneButton.target = self
        doneButton.action = #selector(done)
        doneButton.enabled = true
        doneButton.keyEquivalent = "\r" // busy if button is added into a view
        return doneButton
    }()
    lazy var cancelButton: BLKGUI.Button = {
        let cancelButton = BLKGUI.Button(frame: self.cancelButtonRect)
        cancelButton.title = NSLocalizedString("Cancel", comment: "")
        cancelButton.target = self
        cancelButton.action = #selector(cancel)
        cancelButton.keyEquivalent = "\u{1B}"
        return cancelButton
    }()
    
    var message: String {
        get { return messageLabel.stringValue }
        set(message) {
            let size = message.sizeWithAttributes([NSFontAttributeName: messageLabel.font!])
            messageLabel.alignment = size.width > (messageLabelRect.size.width - 20.0) ? .Left : .Center
            messageLabel.stringValue = message
        }
    }
    
    var text: String {
        get { return textLabel.stringValue }
        set(text) { textLabel.stringValue = text }
    }
    
    init(frame: NSRect, prompt: String) {
        super.init(frame: frame)
        message = prompt
        addSubviews(messageLabel, textLabel, doneButton, cancelButton)
        autoresizingMask = [.ViewMinXMargin, .ViewMaxXMargin, .ViewMinYMargin, .ViewMaxYMargin]
    }
    
    required init(coder: NSCoder) {
        fatalError("NSCoding not supported")
    }
    
    // MARK: Rects

    let margin = NSMakePoint(36.0, 32.0)
    let labelWidth = 85.0
    let buttonSize = NSMakeSize(105.0, 24.0)
    let buttonMargin: CGFloat = 15.0
    let textFont = NSFont.systemFontOfSize(16)
    
    var messageLabelRect: NSRect {
        let size = NSSize(width: bounds.size.width - margin.x * 2, height: 36.0)
        let origin = NSPoint(x: margin.x, y: bounds.size.height - size.height - margin.y)
        return NSRect(origin: origin, size: size)
    }
    
    var textLabelRect: NSRect {
        let size = NSSize(
            width: bounds.size.width - margin.x * 2,
            height: (bounds.size.height - messageLabelRect.size.height - buttonSize.height) - margin.y * 4)
        let origin = NSPoint(x: margin.x, y: messageLabelRect.origin.y - size.height)
        return NSRect(origin: origin, size: size)
    }
    
    var doneButtonRect: NSRect {
        let size = buttonSize
        let origin = NSPoint(
            x: ((bounds.size.width - (size.width * 2 + buttonMargin)) / 2) + size.width + buttonMargin,
            y: margin.y)
        return NSRect(origin: origin, size: size)
    }
    
    var cancelButtonRect: NSRect {
        let size = buttonSize
        let origin = NSPoint(
            x: (bounds.size.width - (size.width * 2 + buttonMargin)) / 2,
            y: margin.y)
        return NSRect(origin: origin, size: size)
    }
    
    // MARK: Delegate
    
    override func controlTextDidChange(_: NSNotification) {
        doneButton.enabled = !textLabel.stringValue.isEmpty
    }
}